### ISSUE TITLE
non-async handlers can be processed sequentially even if concurrency limit is high

### ISSUE THREAD

**Issue by ramonsmits** [Particular/NServiceBus#6797]
#### Is your feature related to a problem? Please describe.

Non-async handlers can be processed sequentially even if the concurrency limit is high.

This has a negative impact on performance which not all users are able to detect or understand.

#### Describe the requested feature

One or more of the following:

- Instead of relying on our users to add Task.Run in their non-async handlers we do this by default in our pipeline
- Add some detection logic that detects that no concurrency is happening
- Have a handler analyzer that detects that no async API's are invoked and suggests to wrap the code in a Task.Run

**Comment by timbussmann** [comment]
@ramonsmits are you thinking about cases where there are multiple handlers for the same message or in situations where the pump is actually pushing multiple messages in parallel? Do you have an example in mind?

**Comment by ramonsmits** [comment]
@timbussmann 

> are you thinking about cases where there are multiple handlers for the same message

no

> in situations where the pump is actually pushing multiple messages in parallel?

Not sure what you mean with this question. If the handlers isn't pretty much immediate going to hit an IO boundary the next message isn't processed in parallel. This can only be achieved via Task.Run in the handlers (or, ideally even much earlier in the pipeline) but I've seen now many times that users do not do that. Meaning, any processing of non-async handler implementations run fully sequential.

**Comment by timbussmann** [comment]
> If the handlers isn't pretty much immediate going to hit an IO boundary the next message isn't processed in parallel.

parallelization is managed by the message pump that pushes messages through the pipeline. To finish a pipeline, the message handler will be fully awaited, regardless if it does async IO or not. It is up to the message pump to use Task.Run if it makes sense for parallelization or if it is using some other approach instead but it doesn't matter how the handler is implemented.

**Comment by ramonsmits** [comment]
> parallelization is managed by the message pump that pushes messages through the pipeline. To finish a pipeline, the message handler will be fully awaited, regardless if it does async IO or not. It is up to the message pump to use Task.Run if it makes sense for parallelization or if it is using some other approach instead but it doesn't matter how the handler is implemented.

I'm not sure what that statement means. We don't have any guidance around transport design that state that messages must be ran in parallel using Task.Run or Yield.

At the moment the only guaranteed method is:

1. Doing async IO in handlers
2. Explicitly use Task.Run in handlers

**Comment by danielmarbach** [comment]
It is true that currently we have no strict enforcement in the transport tests. That being said the pump with it's concurrency settings and our public doc plus acceptance tests running multiple messages in parallel as well as some tests explicitly limiting the concurrency are enough of a safety net. 

see also

> Limit maximum concurrency so that no more messages than the specified value are ever processed at the same time. 

https://docs.particular.net/nservicebus/operations/tuning

By design it is up to the transport to decide whether the pipeline invocations need a preceding yield, task.run or the invocations naturally come from a library that already takes care of it to fulfill the concurrency limits. That's all implementation detail of a transport interacting with a specific transport driver

**Comment by ramonsmits** [comment]
The point it that we aren't ensuring handlers are executed concurrently when the concurrency limit is larger than 1. The tuning section is about the opposite, how to guarantee sequential processing, not how to guarantee concurrent execution. As your saying that isn't strictly enforced by transports while a (majority?) subset of users likely assume that is the default. That actually is related to a comment by NServiceBus veteran @SimonCropp stating that message processing was blocked.

**Comment by timbussmann** [comment]
FYI here's a set of transport tests that actually validate that the pipeline is executed in parallel using the configured concurrency settings: https://github.com/Particular/NServiceBus/pull/6851

**Comment by ramonsmits** [comment]
@timbussmann As this is a new PR does this mean this will only be enforced on v9?

**Comment by timbussmann** [comment]
@ramonsmits yes it's only validated starting from v9, but since there haven't been any significant transport changes, I'd say it's pretty safe to assume that these tests would pass on other major versions too.

**Comment by timbussmann** [comment]
sorry, misclicked. I think this could be closed though?

**Comment by ramonsmits** [comment]
Based on the existence of https://github.com/Particular/NServiceBus/pull/6851 in v9+ yes
